business/namespaces.go

```diff
@@ -129,12 +129,98 @@ func (in *NamespaceService) GetNamespaces(ctx context.Context) ([]models.Namespa
 		labelSelectorExcludeValue = excludeLabelList[1]
 	}
 
+	namespaces := []models.Namespace{}
+
+	var clusterNames []string
+	for c := range in.userClients {
+		clusterNames = append(clusterNames, c)
+	}
+
+	wg := &sync.WaitGroup{}
+	type result struct {
+		cluster string
+		ns      []models.Namespace
+		err     error
+	}
+	resultsCh := make(chan result)
+
+	go func() {
+		for _, cluster := range clusterNames {
+			wg.Add(1)
+			go func(c string) {
+				defer wg.Done()
+				list, error := in.GetNamespacesByCluster(c)
+				if error != nil {
+					log.Errorf("Error fetching Namespaces per cluster %s: %s", c, error)
+					resultsCh <- result{cluster: c, ns: nil, err: error}
+				} else {
+					resultsCh <- result{cluster: c, ns: list, err: nil}
+				}
+			}(cluster)
+		}
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	// Combine namespace data
+	for result := range resultsCh {
+		if result.err != nil {
+			log.Errorf("Error fetching Namespaces per cluster %s: %s", result.cluster, result.err)
+			if result.cluster == kubernetes.HomeClusterName {
+				return nil, result.err
+			}
+		}
+		namespaces = append(namespaces, result.ns...)
+	}
+
+	resultns := namespaces
+
+	// exclude namespaces that are:
+	// 1. to be filtered out via the exclude list
+	// 2. to be filtered out via the label selector
+	// Note that the control plane namespace is never excluded
+	excludes := configObject.API.Namespaces.Exclude
+	if len(excludes) > 0 || labelSelectorExclude != "" {
+		resultns = []models.Namespace{}
+	NAMESPACES:
+		for _, namespace := range namespaces {
+			if namespace.Name != configObject.IstioNamespace {
+				if len(excludes) > 0 {
+					for _, excludePattern := range excludes {
+						if match, _ := regexp.MatchString(excludePattern, namespace.Name); match {
+							continue NAMESPACES
+						}
+					}
+				}
+				if labelSelectorExclude != "" {
+					if namespace.Labels[labelSelectorExcludeName] == labelSelectorExcludeValue {
+						continue NAMESPACES
+					}
+				}
+			}
+			resultns = append(resultns, namespace)
+		}
+	}
+
+	if kialiCache != nil {
+		// Saved per user token
+		kialiCache.SetNamespaces(in.userClients[kubernetes.HomeClusterName].GetToken(), resultns)
```

Comment:
are the tokens common across all clusters? I thought maybe the tokens would be cluster-specific (in which case hardcoding this to the home cluster might not work).

Reply:
Given we're assuming a common OIDC provider, the user tokens will be the same, but the service account tokens will be unique for each cluster.